sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,819 @@
+  %1 = bitcast i32** %ptr to i8*
+
+  %2 = call i8* @llvm.ptr.annotation.p0i8(i8* nonnull %0,
+    i8* getelementptr inbounds ([16 x i8], [16 x i8]* @.str, i64 0, i64 0),
+    i8* getelementptr inbounds ([9 x i8], [9 x i8]* @.str.1, i64 0, i64 0),
+    i32 3,
+    i8* bitcast ({ [9 x i8]*, i8*, [9 x i8]*, i32 }* @.args to i8*))
+
+  %3 = bitcast i8* %2 to i32**
+  store i32* %ptr, i32** %3
+  ret void
+}
+```
+
+The front-end encodes the properties from the C++ attribute
+`[[__sycl_detail__::add_ir_annotations_member()]]` into the
+`@llvm.ptr.annotation` call as follows:
+
+* The first parameter to `@llvm.ptr.annotation` is the pointer to annotate (as
+  with any call to this intrinsic).
+* The second parameter is the literal string `"sycl-properties"`.
+* The third parameter is the name of the source file (as with any call to this
+  intrinsic).
+* The fourth parameter is the line number (as with any call to this intrinsic).
+* The fifth parameter is a metadata tuple with information about all of the
+  properties.  The first element of the tuple is a string literal with the name
+  of the first property.  The second element is the value of the first
+  property.  The third element is a string literal with the name of the second
+  property, etc.  Since each property has exactly one value, this tuple has an
+  even number of elements.
+
+**NOTE**: Calls to the `@llvm.ptr.annotation` intrinsic function are known to
+disable many clang optimizations.  As a result, properties added to a
+non-global variable will likely result in LLVM IR (and SPIR-V) that is not well
+optimized.  This puts more pressure on the SPIR-V consumer (e.g. JIT compiler)
+to perform these optimizations.
+
+
+## Property representation in C++ attributes and in IR
+
+As noted above, there are several C++ attributes that convey property names and
+values to the front-end:
+
+* `[[__sycl_detail__::add_ir_attributes_global_variable()]]`
+* `[[__sycl_detail__::add_ir_attributes_kernel_parameter()]]`
+* `[[__sycl_detail__::add_ir_attributes_function()]]`
+* `[[__sycl_detail__::add_ir_annotations_member()]]`
+
+All of these attributes take a parameter list with the same format.  There are
+always an even number of parameters, where the first half are the property
+names and the second half are the property values.  (This assumes that the
+initial optional parameter is not passed.  See below for a description of this
+optional parameter.) The property name is always a string literal or a
+`constexpr char *` expression.  By convention, property names normally start
+with the prefix `"sycl-"` in order to avoid collision with non-SYCL IR
+attributes, but this is not a strict requirement.
+
+The property value can be a literal or `constexpr` expression of the following
+types:
+
+* `const char *`.
+* An integer type.
+* A floating point type.
+* A boolean type.
+* A character type.
+* An enumeration type.
+* `nullptr_t` (reserved for the case when a property has no value).
```

Comment:
This should be possible to do, but I am not sure why we need it when it will be converted to an empty string anyway. Is there a reason or is it because it is more aesthetically pleasing than `""`?

Reply:
The conversion to the empty string only happens for the first three C++ attributes, not for `[[__sycl_detail__::add_ir_annotations_member()]]`.  I thought it would be nice if the headers passed `nullptr` consistently when the parameter has no value.

It did also seem more aesthetically pleasing to pass `nullptr` as a "no-value" sentinel rather than the empty string.  Does this cause a particular problem with the implementation?